### PR TITLE
[draft] bug for MoE distributed parallelism

### PR DESCRIPTION
**Pull request by realAsma**
## What does this PR do?

**Type of change:** ? 

**Overview:** ?

## Usage

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

File: modelopt/torch/quantization/model_calib.py [hunk 1]
```diff
@@ -81,26 +85,35 @@ def max_calibrate(model: nn.Module, forward_loop: ForwardLoop | None = None, dis
...
     for name, module in model.named_modules():
         if isinstance(module, QuantModule):
             for child in module.children():
-                if isinstance(child, (TensorQuantizer, SequentialQuantizer)):
-                    sync_quantizer_amax_across_dp_ep(child, module.parallel_state)
+                if not isinstance(child, (TensorQuantizer, SequentialQuantizer)):
+                    continue
+                sync_quantizer_amax_across_dp_ep(
+                    child, module.parallel_state, get_module_device(module)
+                )
```
Comment (realAsma) [inline modelopt/torch/quantization/model_calib.py:116]: could you please test if all MoE quantizers have amax after this line (locally)?
```
if `experts` in name and "weight_quantizer` in name:
     assert child.amax is not None
```

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py [hunk 2]
```diff
@@ -1155,19 +1154,62 @@ def set_from_modelopt_state(self, modelopt_state, properties_only: bool = False)
...
+            synced_amax = DistributedProcessGroup.get_dist_syncd_obj(
+                local_amax_cpu,
+                parallel_group,
+                reduce_amaxs,
+            )
+
+            if synced_amax is not None:
+                # Move to target device
+                if target_device is not None:
+                    synced_amax = synced_amax.to(target_device)
```
Comment (jenchen13) [inline modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1227]: need to add 
`synced_amax = synced_amax.clone().detach()`
otherwise the sharding metadata of `global_offset=(0, 0) on all ranks` will be kept during save checkpoint

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py [hunk 3]
```diff
@@ -1155,19 +1154,62 @@ def set_from_modelopt_state(self, modelopt_state, properties_only: bool = False)
...
+        def reduce_amaxs(amax_list: list) -> torch.Tensor | None:
+            """Reduce amax values across ranks by taking element-wise maximum."""
+            valid_amaxs = [a for a in amax_list if a is not None]
+            if not valid_amaxs:
+                return None
+
+            # Iterative max handles both scalar and tensor amax values
+            result = valid_amaxs[0]
+            for amax in valid_amaxs[1:]:
+                result = torch.maximum(result, amax)
```
Comment (jenchen13) [inline modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1211]: what happens if this line is comparing a scalar vs a tensor? how does it determine the max?

Comment (realAsma) [reply modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1211]: see https://docs.pytorch.org/docs/stable/generated/torch.maximum.html 

it simply performs element wise maximum -> the shape does not matter as long as both are pytorch tensors (including scalar tensors)

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py [hunk 4]
```diff
@@ -1155,19 +1154,62 @@ def set_from_modelopt_state(self, modelopt_state, properties_only: bool = False)
...
-                dist.all_reduce(self._amax, op=dist.ReduceOp.MAX, group=parallel_group.group)
-            except RuntimeError as e:
-                # This error happens if the distributed backend is using GPU and
-                # the tensor is not on GPU (or vice versa).
-                warnings.warn(
-                    f"Failed to synchronize amax: {e}, probably because the tensor is on a device which is not"
-                    "supported by the current distributed backend. This warning can be ignored"
-                    "if happening during modelopt restore."
-                )
+    def sync_amax_across_distributed_group(
```
Comment (realAsma) [inline modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1180]: the current sync_amax_across_distributed_group moves the amax to cpu -> this is to accommodate the case were some amaxs are None and some are tensors. However this happens typically only for MoEs.
so can we do the old method of sync for non MoEs:
```
 dist.all_reduce(self._amax, op=dist.ReduceOp.MAX, group=parallel_group.group)
```

and the sync as object via CPU only for MoEs?

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py (hunk 2 above)
Comment (realAsma) [reply modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1227]: Good catch, I am hoping you could take over the PR and address this

Comment (jenchen13) [reply modelopt/torch/quantization/nn/modules/tensor_quantizer.py:1227]: added below
